File: src/lib/contracts/protocol.ts
import { z } from "zod";
import { uuidSchema } from "./common";
import { benchmarkPlanSchema } from "./plan";

export const MAX_FRAME_SIZE_BYTES = 1048576; // 1 MiB

export const protocolMessageTypeSchema = z.enum([
  "HELLO",
  "PAIR_REQUEST",
  "PAIR_RESULT",
  "REQUEST",
  "RESPONSE",
  "PREPARE",
  "PREPARE_RESULT",
  "READY",
  "START",
  "STARTED",
  "SAMPLE",
  "ENGINE_DONE",
  "ENGINE_FAILED",
  "SESSION_RESULT",
  "SESSION_ACK",
  "CANCEL",
  "CANCEL_ACK",
  "HEARTBEAT",
  "ERROR",
  "BYE",
]);
export type ProtocolMessageType = z.infer<typeof protocolMessageTypeSchema>;

export const protocolEnvelopeSchema = z.object({
  type: protocolMessageTypeSchema,
  id: uuidSchema,
  sessionId: uuidSchema.nullable(),
  ts: z.string().datetime({ message: "ts debe ser ISO 8601 UTC" }),
  inReplyTo: uuidSchema.nullable().optional(),
  payload: z.record(z.string(), z.unknown()),
});

export type ProtocolEnvelope = z.infer<typeof protocolEnvelopeSchema>;

// Payloads específicos
export const helloPayloadSchema = z.object({
  protocolVersion: z.number().int().min(1),
  protocolMin: z.number().int().min(1),
  appVersion: z.string().min(1),
  instanceId: uuidSchema,
  displayName: z.string().min(1).max(48),
  platform: z.string(),
  isBusy: z.boolean(),
});
export type HelloPayload = z.infer<typeof helloPayloadSchema>;

export const pairRequestPayloadSchema = z.object({
  pairingCode: z
    .string()
    .regex(/^\d{6}$/, "El código de emparejamiento debe tener exactamente 6 dígitos"),
});
export type PairRequestPayload = z.infer<typeof pairRequestPayloadSchema>;

export const pairResultPayloadSchema = z.object({
  accepted: z.boolean(),
  reason: z.string().nullable().optional(),
});
export type PairResultPayload = z.infer<typeof pairResultPayloadSchema>;

export const requestPayloadSchema = z.object({
  plan: benchmarkPlanSchema,
  estimateSeconds: z.number().int().positive(),
  suggestedInterface: z.string().nullable().optional(),
});
export type RequestPayload = z.infer<typeof requestPayloadSchema>;

export const responsePayloadSchema = z.object({
  accepted: z.boolean(),
  reason: z.string().nullable().optional(),
  interfaceName: z.string().nullable().optional(),
});
export type ResponsePayload = z.infer<typeof responsePayloadSchema>;

export const cancelPayloadSchema = z.object({
  reason: z.string().min(1),
  code: z.string().nullable().optional(),
});
export type CancelPayload = z.infer<typeof cancelPayloadSchema>;

export const heartbeatPayloadSchema = z.object({
  nonce: z.number().int().nonnegative(),
});
export type HeartbeatPayload = z.infer<typeof heartbeatPayloadSchema>;

File: src-tauri/src/model/protocol.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub const MAX_FRAME_SIZE_BYTES: usize = 1_048_576; // 1 MiB

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ProtocolMessageType {
    Hello,
    PairRequest,
    PairResult,
    Request,
    Response,
    Prepare,
    PrepareResult,
    Ready,
    Start,
    Started,
    Sample,
    EngineDone,
    EngineFailed,
    SessionResult,
    SessionAck,
    Cancel,
    CancelAck,
    Heartbeat,
    Error,
    Bye,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProtocolEnvelope<T> {
    #[serde(rename = "type")]
    pub msg_type: ProtocolMessageType,
    pub id: Uuid,
    pub session_id: Option<Uuid>,
    pub ts: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_reply_to: Option<Uuid>,
    pub payload: T,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HelloPayload {
    pub protocol_version: u32,
    pub protocol_min: u32,
    pub app_version: String,
    pub instance_id: Uuid,
    pub display_name: String,
    pub platform: String,
    pub is_busy: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PairRequestPayload {
    pub pairing_code: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PairResultPayload {
    pub accepted: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RequestPayload {
    pub plan: super::plan::BenchmarkPlan,
    pub estimate_seconds: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub suggested_interface: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResponsePayload {
    pub accepted: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interface_name: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelPayload {
    pub reason: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HeartbeatPayload {
    pub nonce: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreparePayload {
    pub direction: String,
    pub port: u16,
    pub streams: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReadyPayload {
    pub direction: String,
    pub is_ready: bool,
    pub listening_port: u16,
}

/// Negociación de inicio sin sincronización de reloj absoluto (Constitución VII):
/// start_delay_ms: offset relativo en milisegundos desde la recepción del mensaje
/// tolerance_ms: ventana máxima admisible antes de declarar desincronización
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartPayload {
    pub direction: String,
    pub start_delay_ms: u64,
    pub tolerance_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartedPayload {
    pub direction: String,
    pub local_t_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EngineDonePayload {
    pub direction: String,
    pub role: String,
    pub throughput_bps: String,
    pub total_bytes: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionAckPayload {
    pub session_id: Uuid,
    pub persisted: bool,
    pub thresholds_hash: String,
}

/// Framing helpers:
/// 4 bytes big-endian length prefix + UTF-8 payload.
pub fn encode_frame(payload_json: &[u8]) -> Result<Vec<u8>, String> {
    let len = payload_json.len();
    if len > MAX_FRAME_SIZE_BYTES {
        return Err(format!(
            "El tamaño del payload excede el límite máximo de 1 MiB (tamaño: {} bytes)",
            len
        ));
    }
    let mut frame = Vec::with_capacity(4 + len);
    frame.extend_from_slice(&(len as u32).to_be_bytes());
    frame.extend_from_slice(payload_json);
    Ok(frame)
}

pub fn decode_frame_length(header: [u8; 4]) -> usize {
    u32::from_be_bytes(header) as usize
}
